README.md
# opengl-miscellaneous

<!-- badges: start -->
[![Stack-lts](https://github.com/stla/opengl-miscellaneous/actions/workflows/Stack-lts.yml/badge.svg)](https://github.com/stla/opengl-miscellaneous/actions/workflows/Stack-lts.yml)
[![Stack-lts-linux](https://github.com/stla/opengl-miscellaneous/actions/workflows/Stack-lts-linux.yml/badge.svg)](https://github.com/stla/opengl-miscellaneous/actions/workflows/Stack-lts-linux.yml)
[![Stack-nightly](https://github.com/stla/opengl-miscellaneous/actions/workflows/Stack-nightly.yml/badge.svg)](https://github.com/stla/opengl-miscellaneous/actions/workflows/Stack-nightly.yml)
[![Stack-nightly-linux](https://github.com/stla/opengl-miscellaneous/actions/workflows/Stack-nightly-linux.yml/badge.svg)](https://github.com/stla/opengl-miscellaneous/actions/workflows/Stack-nightly-linux.yml)
<!-- badges: end -->


![](gallery/wavyTorus.gif)

![](gallery/gears.gif)

![](gallery/HopfTorus.gif)

![](gallery/HopfCylinder.gif)


CHANGELOG.md
# Changelog for `opengl-miscellaneous`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


src/Gear.hs
module Gear
  (main) 
  where
import           Control.Monad                     (when)
import qualified Data.ByteString                   as B
import           Data.IORef
import qualified Data.Map.Strict                   as M
import           Gear.Data
import           Graphics.Rendering.OpenGL.Capture (capturePPM)
import           Graphics.Rendering.OpenGL.GL
import           Graphics.UI.GLUT                  hiding (shift)
import           Text.Printf
import           Utils.Palettes                    (colorRampSymmetric')

white,black :: Color4 GLfloat
white      = Color4    1    1    1    1
black      = Color4    0    0    0    1

n_u,n_v :: Int
n_u = 250
n_v = 2


display :: IORef GLfloat -> IORef GLfloat -> IORef GLfloat -- rotations
        -> IORef String    -- color palette
        -> IORef Int       -- shift palette
        -> IORef GLdouble  -- parameter n
        -> IORef GLdouble  -- zoom
        -> IORef GLfloat   -- gears rotation
        -> DisplayCallback
display rot1 rot2 rot3 palette shift n zoom angle = do
  clear [ColorBuffer, DepthBuffer]
  n' <- get n
  r1 <- get rot1
  r2 <- get rot2
  r3 <- get rot3
  z <- get zoom
  angle' <- get angle
  (_, size) <- get viewport
  palette' <- get palette
  shift' <- get shift
  let colors = colorRampSymmetric' palette' n_u (shift' `mod` n_u)
      surface = M.toList $ allQuads n_u n_v n'
  loadIdentity
  resize z size
  rotate r1 $ Vector3 1 0 0
  rotate r2 $ Vector3 0 1 0
  rotate r3 $ Vector3 0 0 1
--  rotate (pi/4 :: GLfloat) $ Vector3 0 1 0
  preservingMatrix $ do
    rotate angle' $ Vector3 0 0 1
    renderPrimitive Quads $ mapM_ (drawQuad colors) surface
  preservingMatrix $ do
    translate (Vector3 2.1 0 (0 :: GLfloat))
    rotate (-angle') $ Vector3 0 0 1
    renderPrimitive Quads $ mapM_ (drawQuad colors) surface
  swapBuffers
  where
    drawQuad thecolors ((i,_), quad) = do
      materialDiffuse FrontAndBack $= thecolors !! i
      drawQuad' quad
        where
          drawQuad' ((v1,v2,v3,v4),norm) = do
            normal norm
            vertex v1
            vertex v2
            vertex v3
            vertex v4

resize :: Double -> Size -> IO ()
resize zoom s@(Size w h) = do
  viewport $= (Position 0 0, s)
  matrixMode $= Projection
  loadIdentity
  perspective 45.0 (w'/h') 1.0 100.0
  lookAt (Vertex3 6 0 (-3+zoom)) (Vertex3 1 0 0) (Vector3 0 1 0)
  matrixMode $= Modelview 0
  where
    w' = realToFrac w
    h' = realToFrac h

keyboard :: IORef GLfloat -> IORef GLfloat -> IORef GLfloat -- rotations
         -> IORef Int      -- shift palette
         -> IORef GLdouble -- parameter n
         -> IORef GLdouble -- zoom
         -> IORef Bool     -- animation
         -> KeyboardCallback
keyboard rot1 rot2 rot3 shift n zoom anim c _ =
  case c of
    'a' -> writeIORef anim True
    's' -> shift $~! (+2)
    'b' -> n $~! subtract 0.1
    'n' -> n $~! (+ 0.1)
    'e' -> rot1 $~! subtract 2
    'r' -> rot1 $~! (+2)
    't' -> rot2 $~! subtract 2
    'y' -> rot2 $~! (+2)
    'u' -> rot3 $~! subtract 2
    'i' -> rot3 $~! (+2)
    'm' -> zoom $~! (+1)
    'l' -> zoom $~! subtract 1
    'q' -> leaveMainLoop
    _   -> return ()

idle :: IORef Bool -> IORef GLfloat -> IORef Int -> IdleCallback
idle anim angle snapshots = do
    a <- get anim
    s <- get snapshots
    when a $ do
      when (s < 360) $ do
        let ppm = printf "ppm/gears%04d.ppm" s
        (>>=) capturePPM (B.writeFile ppm)
      angle $~! (+ 1)
      snapshots $~! (+ 1)
    postRedisplay Nothing

menuPalette :: IORef String -> String -> MenuCallback
menuPalette = writeIORef

main :: IO ()
main = do
  _ <- getArgsAndInitialize
  _ <- createWindow "Gears"
  windowSize $= Size 512 512
  initialDisplayMode $= [RGBAMode, DoubleBuffered, WithDepthBuffer]
  clearColor $= white
  materialAmbient FrontAndBack $= black
  materialShininess FrontAndBack $= 95
  materialSpecular FrontAndBack $= white
  lighting $= Enabled
  lightModelTwoSide $= Enabled
  light (Light 0) $= Enabled
  position (Light 0) $= Vertex4 200 0 (-100) 1
  ambient (Light 0) $= black
  diffuse (Light 0) $= white
  specular (Light 0) $= white
  depthFunc $= Just Less
  shadeModel $= Smooth
  n <- newIORef 12.0
  rot1 <- newIORef 0.0
  rot2 <- newIORef 0.0
  rot3 <- newIORef 0.0
  zoom <- newIORef 0.0
  angle <- newIORef 0.0
  palette <- newIORef "viridis"
  shift <- newIORef 0
  anim <- newIORef False
  snapshots <- newIORef 0
  displayCallback $= display rot1 rot2 rot3 palette shift n zoom angle
  reshapeCallback $= Just (resize 0)
  keyboardCallback $= Just (keyboard rot1 rot2 rot3 shift n zoom anim)
  idleCallback $= Just (idle anim angle snapshots)
  putStrLn "*** Gears ***\n\
        \    To quit, press q.\n\
        \    Scene rotation:\n\
        \        e, r, t, y, u, i\n\
        \    Zoom: l, m\n\
        \    Increase/decrease parameter: n, b \n\
        \    Rotate colors: s \n\
        \    Animation: a \n\
        \"
  attachMenu LeftButton
    (Menu [ SubMenu "Color palette"
                    (Menu [ MenuEntry "magma"    (menuPalette palette "magma")
                          , MenuEntry "inferno"  (menuPalette palette "inferno")
                          , MenuEntry "plasma"   (menuPalette palette "plasma")
                          , MenuEntry "viridis"  (menuPalette palette "viridis")
                          , MenuEntry "cviridis" (menuPalette palette "cviridis")
                          ])
          ]
      )
  mainLoop




src/WavyTorus.hs
module WavyTorus
  (main)
  where
import           Control.Monad                     (when)
import qualified Data.ByteString                   as B
import           Data.IORef
import qualified Data.Map.Strict                   as M
import           Graphics.Rendering.OpenGL.Capture (capturePPM)
import           Graphics.Rendering.OpenGL.GL
import           Graphics.UI.GLUT                  hiding (shift)
import           Text.Printf
import           Utils.Palettes                    (colorRampSymmetric')
import           WavyTorus.Data

white,black :: Color4 GLfloat
white      = Color4    1    1    1    1
black      = Color4    0    0    0    1

half_n_u,n_u,n_v :: Int
half_n_u = 250
n_u = 2 * half_n_u
n_v = 100


display :: IORef GLfloat -> IORef GLfloat -> IORef GLfloat -- rotations
        -> IORef String    -- color palette
        -> IORef Int       -- shift palette
        -> IORef GLdouble  -- parameter n
        -> IORef GLdouble  -- zoom
        -> DisplayCallback
display rot1 rot2 rot3 palette shift n zoom = do
  clear [ColorBuffer, DepthBuffer]
  n' <- get n
  r1 <- get rot1
  r2 <- get rot2
  r3 <- get rot3
  z <- get zoom
  (_, size) <- get viewport
  palette' <- get palette
  shift' <- get shift
  let colors = colorRampSymmetric' palette' n_u (shift' `mod` n_u)
      surface = allQuads n_u n_v n'
  loadIdentity
  resize z size
  rotate r1 $ Vector3 1 0 0
  rotate r2 $ Vector3 0 1 0
  rotate r3 $ Vector3 0 0 1
  renderPrimitive Quads $ mapM_ (drawQuad colors) (M.toList surface)
  swapBuffers
  where
    drawQuad thecolors ((i,_), quad) = do
      materialDiffuse FrontAndBack $= thecolors !! i
      drawQuad' quad
        where
          drawQuad' ((v1,v2,v3,v4),norm) = do
            normal norm
            vertex v1
            vertex v2
            vertex v3
            vertex v4

resize :: Double -> Size -> IO ()
resize zoom s@(Size w h) = do
  viewport $= (Position 0 0, s)
  matrixMode $= Projection
  loadIdentity
  perspective 45.0 (w'/h') 1.0 100.0
  lookAt (Vertex3 0 0 (-15+zoom)) (Vertex3 0 0 0) (Vector3 0 1 0)
  matrixMode $= Modelview 0
  where
    w' = realToFrac w
    h' = realToFrac h

keyboard :: IORef GLfloat -> IORef GLfloat -> IORef GLfloat -- rotations
         -> IORef Int      -- shift palette
         -> IORef GLdouble -- parameter n
         -> IORef GLdouble -- zoom
         -> IORef Bool     -- animation
         -> KeyboardCallback
keyboard rot1 rot2 rot3 shift n zoom anim c _ =
  case c of
    'a' -> writeIORef anim True
    's' -> shift $~! (+2)
    'b' -> n $~! subtract 0.1
    'n' -> n $~! (+ 0.1)
    'e' -> rot1 $~! subtract 2
    'r' -> rot1 $~! (+2)
    't' -> rot2 $~! subtract 2
    'y' -> rot2 $~! (+2)
    'u' -> rot3 $~! subtract 2
    'i' -> rot3 $~! (+2)
    'm' -> zoom $~! (+1)
    'l' -> zoom $~! subtract 1
    'q' -> leaveMainLoop
    _   -> return ()

idle :: IORef Bool -> IORef Int -> IORef Int -> IdleCallback
idle anim shift snapshots = do
    a <- get anim
    s <- get snapshots
    when a $ do
      when (s < half_n_u) $ do
        let ppm = printf "ppm/wavytorus%04d.ppm" s
        (>>=) capturePPM (B.writeFile ppm)
      shift $~! (+ 2)
      snapshots $~! (+ 1)
    postRedisplay Nothing

menuPalette :: IORef String -> String -> MenuCallback
menuPalette = writeIORef

main :: IO ()
main = do
  _ <- getArgsAndInitialize
  _ <- createWindow "Wavy Torus"
  windowSize $= Size 512 512
  initialDisplayMode $= [RGBAMode, DoubleBuffered, WithDepthBuffer]
  clearColor $= white
  materialAmbient FrontAndBack $= black
  materialShininess FrontAndBack $= 95
  materialSpecular FrontAndBack $= white
  lighting $= Enabled
  lightModelTwoSide $= Enabled
  light (Light 0) $= Enabled
  position (Light 0) $= Vertex4 0 0 (-100) 1
  ambient (Light 0) $= black
  diffuse (Light 0) $= white
  specular (Light 0) $= white
  depthFunc $= Just Less
  shadeModel $= Smooth
  n <- newIORef 5.0
  rot1 <- newIORef 0.0
  rot2 <- newIORef 0.0
  rot3 <- newIORef 0.0
  zoom <- newIORef 0.0
  palette <- newIORef "viridis"
  shift <- newIORef 0
  anim <- newIORef False
  snapshots <- newIORef 0
  displayCallback $= display rot1 rot2 rot3 palette shift n zoom
  reshapeCallback $= Just (resize 0)
  keyboardCallback $= Just (keyboard rot1 rot2 rot3 shift n zoom anim)
  idleCallback $= Just (idle anim shift snapshots)
  putStrLn "*** Wavy Torus ***\n\
        \    To quit, press q.\n\
        \    Scene rotation:\n\
        \        e, r, t, y, u, i\n\
        \    Zoom: l, m\n\
        \    Increase/decrease parameter: n, b \n\
        \    Rotate colors: s \n\
        \    Animation: a \n\
        \"
  attachMenu LeftButton
    (Menu [ SubMenu "Color palette"
                    (Menu [ MenuEntry "magma"    (menuPalette palette "magma")
                          , MenuEntry "inferno"  (menuPalette palette "inferno")
                          , MenuEntry "plasma"   (menuPalette palette "plasma")
                          , MenuEntry "viridis"  (menuPalette palette "viridis")
                          , MenuEntry "cviridis" (menuPalette palette "cviridis")
                          ])
          ]
      )
  mainLoop




src/Utils/Hopf.hs
module Utils.Hopf
  (stereoCircHinv'', stereoCircHinv)
  where

hopfinverse :: Floating a => a -> a -> a -> a -> (a,a,a,a)
hopfinverse q0 q1 q2 t =
  let h = 1 / sqrt (2 * (1+q0)) in
  (h * (cos t * q2 + sin t * q1),
   h * (cos t * q1 - sin t * q2),
   h * sin t * (1+q0),
   h * cos t * (1+q0))

-- hopfinverse' :: (Num a, Floating a) => a -> a -> a -> (a,a,a,a)
-- hopfinverse' theta phi =
--   hopfinverse (cos theta * cos phi) (sin theta * cos phi) (sin phi) 

stereoProj :: Floating a => (a,a,a,a) -> (a,a,a)
stereoProj (x0,x1,x2,x3) =
  let h = 1/(1-x3) in
  (h * x0, h * x1, h * x2)

stereoCircHinv :: Floating a => a -> a -> a -> (a, a, a)
stereoCircHinv theta phi xi =
  stereoProj (hopfinverse q0 q1 q2 xi)
  where
    q0 = cos theta * cos phi
    q1 = sin theta * cos phi
    q2 = sin phi

-- stereoCircHinv' :: (Num a, Floating a) => a -> a -> a -> (a, a, a)
-- stereoCircHinv' theta phi xi =
--   stereoProj (hopfinverse q0 q1 q2 xi)
--   where
--     q0 = cos theta * sin phi
--     q1 = sin theta * sin phi
--     q2 = cos phi

stereoCircHinv'' :: Floating a => a -> a -> a -> a -> (a, a, a)
stereoCircHinv'' q0 q1 q2 xi =
  stereoProj (hopfinverse q0 q1 q2 xi)


src/HopfCylinder.hs
module HopfCylinder
  (main)
  where
import           Control.Monad                     (forM_, when)
import qualified Data.ByteString                   as B
import           Data.IORef
import           Data.Tuple.Extra                  (fst3, snd3, thd3)
import           Graphics.Rendering.OpenGL.Capture (capturePPM)
import           Graphics.Rendering.OpenGL.GL
import           Graphics.UI.GLUT
import           Linear                            (V3 (..))
import           Text.Printf
import           Utils.Hopf
import           Utils.TorusTransformationMatrix

type Point = (GLfloat,GLfloat,GLfloat)

pointToV3 :: Point -> V3 GLfloat
pointToV3 (x,y,z) = V3 x y z

tripletOnCircle :: GLfloat -> (Point, Point, Point)
tripletOnCircle theta =
  let phi = 0.25 * atan (1 / cos (theta - pi/2)) in
  ( stereoCircHinv theta phi 0
  , stereoCircHinv theta phi 2
  , stereoCircHinv theta phi 4)

myTriplets :: Int -> [(Point, Point, Point)]
myTriplets n = map tripletOnCircle theta_ -- n is number of circles
  where
    theta_ = [2 * pi * frac i n | i <- [0 .. n-1]]
    frac :: Int -> Int -> GLfloat
    frac p q = realToFrac p / realToFrac q

data Context = Context
    {
      contextRot1 :: IORef GLfloat
    , contextRot2 :: IORef GLfloat
    , contextRot3 :: IORef GLfloat
    , contextZoom :: IORef Double
    }

white,black,blue :: Color4 GLfloat
white      = Color4    1    1    1    1
black      = Color4    0    0    0    1
blue       = Color4    0    0    1    1

tmatsAndRadii :: Int -> [([GLfloat], GLdouble)]
tmatsAndRadii n = map (\triplet -> torusTransformationMatrix
                                     (pointToV3 $ fst3 triplet)
                                     (pointToV3 $ snd3 triplet)
                                     (pointToV3 $ thd3 triplet))
                      (myTriplets n)

display :: Context -> IORef GLfloat -> IORef Int -> DisplayCallback
display context alpha n = do
  clear [ColorBuffer, DepthBuffer]
  r1 <- get (contextRot1 context)
  r2 <- get (contextRot2 context)
  r3 <- get (contextRot3 context)
  zoom <- get (contextZoom context)
  alpha' <- get alpha
  n' <- get n
  (_, size) <- get viewport
  loadIdentity
  resize zoom size
  rotate r1 $ Vector3 1 0 0
  rotate r2 $ Vector3 0 1 0
  rotate r3 $ Vector3 0 0 1
  rotate alpha' $ Vector3 1 1 1
  forM_ (tmatsAndRadii n') $ \tmatAndRadius ->
    preservingMatrix $ do
      tmat <- newMatrix RowMajor (fst tmatAndRadius) :: IO (GLmatrix GLfloat)
      multMatrix tmat
      materialDiffuse Front $= blue
      renderObject Solid $ Torus 0.2 (snd tmatAndRadius) 30 30
  swapBuffers

resize :: Double -> Size -> IO ()
resize zoom s@(Size w h) = do
  viewport $= (Position 0 0, s)
  matrixMode $= Projection
  loadIdentity
  perspective 45.0 (w'/h') 1.0 100.0
  lookAt (Vertex3 0 0 (-35+zoom)) (Vertex3 0 0 0) (Vector3 0 1 0)
  matrixMode $= Modelview 0
  where
    w' = realToFrac w
    h' = realToFrac h

keyboard :: IORef GLfloat -> IORef GLfloat -> IORef GLfloat -- rotations
         -> IORef GLdouble -- zoom
         -> IORef Bool -- animation
         -> IORef Int -- number of circles
         -> KeyboardCallback
keyboard rot1 rot2 rot3 zoom anim n c _ = do
  case c of
    'a' -> writeIORef anim True
    'h' -> n $~! (+50)
    'n' -> n $~! \x -> if x>100 then x-50 else x
    'e' -> rot1 $~! subtract 2
    'r' -> rot1 $~! (+2)
    't' -> rot2 $~! subtract 2
    'y' -> rot2 $~! (+2)
    'u' -> rot3 $~! subtract 2
    'i' -> rot3 $~! (+2)
    'm' -> zoom $~! (+1)
    'l' -> zoom $~! subtract 1
    'q' -> leaveMainLoop
    _   -> return ()
  postRedisplay Nothing

idle :: IORef Bool -> IORef Int -> IORef GLfloat -> IdleCallback
idle anim snapshots alpha = do
    a <- get anim
    s <- get snapshots
    when a $ do
      when (s < 360) $ do
        let ppm = printf "ppm/hcylinder%04d.ppm" s
        (>>=) capturePPM (B.writeFile ppm)
      snapshots $~! (+1)
      alpha $~! (+1)
      postRedisplay Nothing
    return ()

main :: IO ()
main = do
  _ <- getArgsAndInitialize
  _ <- createWindow "Hopf cylinder"
  windowSize $= Size 512 512
  initialDisplayMode $= [RGBAMode, DoubleBuffered, WithDepthBuffer]
  clearColor $= white
  materialAmbient Front $= black
  lighting $= Enabled
  light (Light 0) $= Enabled
  position (Light 0) $= Vertex4 0 0 (-100) 1
  ambient (Light 0) $= black
  diffuse (Light 0) $= white
  specular (Light 0) $= black
  depthFunc $= Just Less
  shadeModel $= Smooth
  rot1 <- newIORef 0.0
  rot2 <- newIORef 0.0
  rot3 <- newIORef 0.0
  zoom <- newIORef 0.0
  alpha <- newIORef 0.0
  n <- newIORef 150
  anim <- newIORef False
  snapshots <- newIORef 0
  displayCallback $= display Context { contextRot1 = rot1,
                                       contextRot2 = rot2,
                                       contextRot3 = rot3,
                                       contextZoom = zoom }
                             alpha n
  reshapeCallback $= Just (resize 0)
  keyboardCallback $= Just (keyboard rot1 rot2 rot3 zoom anim n)
  idleCallback $= Just (idle anim snapshots alpha)
  putStrLn "*** Hopf torus ***\n\
        \    To quit, press q.\n\
        \    Scene rotation:\n\
        \        e, r, t, y, u, i\n\
        \    Zoom: l, m\n\
        \    Increase/decrease number of circles: h,n\n\
        \    Animation: a\n\
        \"
  mainLoop


src/Utils/TorusTransformationMatrix.hs
module Utils.TorusTransformationMatrix
  (torusTransformationMatrix)
  where
import           Data.Foldable                (toList)
import           Graphics.Rendering.OpenGL.GL (GLdouble)
import           Linear

-- | the plane passing by three points
plane3pts :: Num a => V3 a -> V3 a -> V3 a -> (V3 a, a)
plane3pts p1 p2 p3 = (V3 xcoef ycoef zcoef, offset)
  where
    V3 p1x p1y p1z = p1
    V3 p2x p2y p2z = p2
    V3 p3x p3y p3z = p3
    xcoef = (p1y-p2y)*(p2z-p3z)-(p1z-p2z)*(p2y-p3y)
    ycoef = (p1z-p2z)*(p2x-p3x)-(p1x-p2x)*(p2z-p3z)
    zcoef = (p1x-p2x)*(p2y-p3y)-(p1y-p2y)*(p2x-p3x)
    offset = p1x*xcoef + p1y*ycoef + p1z*zcoef

-- | plane given a point and a normal
plane1ptnormal :: Num a => V3 a -> V3 a -> (V3 a, a)
plane1ptnormal p normal = (normal, p `dot` normal)

-- | circumcenter and circumradius given three points
circleCenterRadius :: Floating a => V3 a -> V3 a -> V3 a
                   -> ((V3 a, a), V3 a)
circleCenterRadius p1 p2 p3 = ((center, radius), coefs1)
  where
    p12 = (p1 ^+^ p2) ^/ 2
    p23 = (p2 ^+^ p3) ^/ 2
    v12 = p2 ^-^ p1
    v23 = p3 ^-^ p2
    (coefs1, offset1) = plane3pts p1 p2 p3
    (coefs2, offset2) = plane1ptnormal p12 v12
    (coefs3, offset3) = plane1ptnormal p23 v23
    a = V3 coefs1 coefs2 coefs3
    b = V3 offset1 offset2 offset3
    center = inv33 a !* b
    op1 = p1 ^-^ center
    radius = norm op1

-- | the transformation matrix for the torus passing through three points
torusTransformationMatrix :: (Real a, Floating a) => V3 a -> V3 a -> V3 a
                     -> ([a], GLdouble)
torusTransformationMatrix p1 p2 p3 =
  (concatMap toList (toList (mkTransformationMat m center)), radius')
  where
    ((center, radius), plane) = circleCenterRadius p1 p2 p3
    radius' = realToFrac radius
    V3 a b c = plane
    measure = sqrt (a*a + b*b + c*c)
    a' = a / measure
    b' = b / measure
    c' = c / measure
    n = V3 a' b' c'
    s = sqrt (a'*a' + b'*b') -- TODO: case a=b=0
    a'' = a' / s
    b'' = b' / s
    u = V3 b'' (-a'') 0
    v = cross n u
    m = transpose $ V3 u v n


src/HopfTorus.hs
module HopfTorus
  (main)
  where
import           Control.Monad                     (forM_, when)
import qualified Data.ByteString                   as B
import           Data.IORef
import           Data.Tuple.Extra                  (fst3, snd3, thd3)
import           Graphics.Rendering.OpenGL.Capture (capturePPM)
import           Graphics.Rendering.OpenGL.GL
import           Graphics.UI.GLUT
import           Linear                            (V3 (..))
import           Text.Printf
import           Utils.Hopf
import           Utils.TorusTransformationMatrix

type Point = (GLfloat,GLfloat,GLfloat)

pointToV3 :: Point -> V3 GLfloat
pointToV3 (x,y,z) = V3 x y z

tripletOnCircle :: Int -> GLfloat -> (Point, Point, Point)
tripletOnCircle m u = -- m is the number of lobes
  ( stereoCircHinv'' x y z 0
  , stereoCircHinv'' x y z 2
  , stereoCircHinv'' x y z 4)
  where
    a = 0.44
    nlobes = fromIntegral m
    b = pi/2 - (pi/2 - a)*cos(u*nlobes)
    c = u + a*sin(2*u*nlobes)
    x = cos b
    sinb = sin b
    y = sinb * cos c
    z = sinb * sin c

myTriplets :: Int -> Int -> [(Point, Point, Point)]
myTriplets m n = map (tripletOnCircle m) u_ -- n: number of circles
  where
    u_ = [-pi + 2 * pi * frac i n | i <- [0 .. n-1]]
    frac :: Int -> Int -> GLfloat
    frac p q = realToFrac p / realToFrac q

data Context = Context
    {
      contextRot1 :: IORef GLfloat
    , contextRot2 :: IORef GLfloat
    , contextRot3 :: IORef GLfloat
    , contextZoom :: IORef Double
    }

white,black,blue :: Color4 GLfloat
white      = Color4    1    1    1    1
black      = Color4    0    0    0    1
blue       = Color4    0    0    1    1

tmatsAndRadii :: Int -> Int -> [([GLfloat], GLdouble)]
tmatsAndRadii m n = map (\triplet -> torusTransformationMatrix
                                     (pointToV3 $ fst3 triplet)
                                     (pointToV3 $ snd3 triplet)
                                     (pointToV3 $ thd3 triplet))
                        (myTriplets m n)

display :: Context -> IORef GLfloat -> IORef Int -> IORef Int -> DisplayCallback
display context alpha m n = do
  clear [ColorBuffer, DepthBuffer]
  r1 <- get (contextRot1 context)
  r2 <- get (contextRot2 context)
  r3 <- get (contextRot3 context)
  zoom <- get (contextZoom context)
  alpha' <- get alpha
  m' <- get m
  n' <- get n
  (_, size) <- get viewport
  loadIdentity
  resize zoom size
  rotate r1 $ Vector3 1 0 0
  rotate r2 $ Vector3 0 1 0
  rotate r3 $ Vector3 0 0 1
  rotate alpha' $ Vector3 1 1 1
  forM_ (tmatsAndRadii m' n') $ \tmatAndRadius ->
    preservingMatrix $ do
      tmat <- newMatrix RowMajor (fst tmatAndRadius) :: IO (GLmatrix GLfloat)
      multMatrix tmat
      materialDiffuse Front $= blue
      renderObject Solid $ Torus 0.2 (snd tmatAndRadius) 30 30
  swapBuffers

resize :: Double -> Size -> IO ()
resize zoom s@(Size w h) = do
  viewport $= (Position 0 0, s)
  matrixMode $= Projection
  loadIdentity
  perspective 45.0 (w'/h') 1.0 100.0
  lookAt (Vertex3 0 0 (-35+zoom)) (Vertex3 0 0 0) (Vector3 0 1 0)
  matrixMode $= Modelview 0
  where
    w' = realToFrac w
    h' = realToFrac h

keyboard :: IORef GLfloat -> IORef GLfloat -> IORef GLfloat -- rotations
         -> IORef GLdouble -- zoom
         -> IORef Bool -- animation
         -> IORef Int -- number of lobes
         -> IORef Int -- number of circles
         -> KeyboardCallback
keyboard rot1 rot2 rot3 zoom anim m n c _ = do
  case c of
    'a' -> writeIORef anim True
    'g' -> m $~! (+1)
    'b' -> m $~! \x -> if x>1 then x-1 else x
    'h' -> n $~! (+50)
    'n' -> n $~! \x -> if x>100 then x-50 else x
    'e' -> rot1 $~! subtract 2
    'r' -> rot1 $~! (+2)
    't' -> rot2 $~! subtract 2
    'y' -> rot2 $~! (+2)
    'u' -> rot3 $~! subtract 2
    'i' -> rot3 $~! (+2)
    'm' -> zoom $~! (+1)
    'l' -> zoom $~! subtract 1
    'q' -> leaveMainLoop
    _   -> return ()
  postRedisplay Nothing

idle :: IORef Bool -> IORef Int -> IORef GLfloat -> IdleCallback
idle anim snapshots alpha = do
    a <- get anim
    s <- get snapshots
    when a $ do
      when (s < 360) $ do
        let ppm = printf "ppm/htorus%04d.ppm" s
        (>>=) capturePPM (B.writeFile ppm)
      snapshots $~! (+1)
      alpha $~! (+1)
      postRedisplay Nothing
    return ()

main :: IO ()
main = do
  _ <- getArgsAndInitialize
  _ <- createWindow "Hopf torus"
  windowSize $= Size 512 512
  initialDisplayMode $= [RGBAMode, DoubleBuffered, WithDepthBuffer]
  clearColor $= white
  materialAmbient Front $= black
  lighting $= Enabled
  light (Light 0) $= Enabled
  position (Light 0) $= Vertex4 0 0 (-100) 1
  ambient (Light 0) $= black
  diffuse (Light 0) $= white
  specular (Light 0) $= black
  depthFunc $= Just Less
  shadeModel $= Smooth
  rot1 <- newIORef 0.0
  rot2 <- newIORef 0.0
  rot3 <- newIORef 0.0
  zoom <- newIORef 0.0
  alpha <- newIORef 0.0
  m <- newIORef 3
  n <- newIORef 150
  anim <- newIORef False
  snapshots <- newIORef 0
  displayCallback $= display Context { contextRot1 = rot1,
                                       contextRot2 = rot2,
                                       contextRot3 = rot3,
                                       contextZoom = zoom }
                             alpha m n
  reshapeCallback $= Just (resize 0)
  keyboardCallback $= Just (keyboard rot1 rot2 rot3 zoom anim m n)
  idleCallback $= Just (idle anim snapshots alpha)
  putStrLn "*** Hopf torus ***\n\
        \    To quit, press q.\n\
        \    Scene rotation:\n\
        \        e, r, t, y, u, i\n\
        \    Zoom: l, m\n\
        \    Increase/decrease number of lobes: g,b\n\
        \    Increase/decrease number of circles: h,n\n\
        \    Animation: a\n\
        \"
  mainLoop
